### PR TITLE
Issue #14966: Update MethodParmPad to support record patterns

### LINKED ISSUE #14966: Add Check Support for Java 21 Record Pattern : MethodParamPad
child of #14890 

I have read check documentation: https://checkstyle.org/checks/whitespace/methodparampad.html#MethodParamPad
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/RecordPatterns.java
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.List;

public class RecordPatterns {
    class A { }
    sealed interface I permits C, D { }
    final class C implements I { }
    final class D implements I { }
    record Pair<T> (T x, T y) { } // violation

    void test() {
        C c = new C();
        D d = new D();

        Pair<I> p2 = new Pair<>(c, d);

        switch (p2) {
            case Pair (I i, C _) -> System.out.println("Pair (C|D, C)");    // expected violation
            case Pair(I i, D j) -> System.out.println("Pair (C|D, D)");
        }

    }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="MethodParamPad"/>
 </module>
</module>
PS D:\CS\test>  java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:8:20: '(' is preceded with whitespace. [MethodParamPad]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

I want a violation on the record pattern

---

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #14966 :

Diff Regression config: https://gist.githubusercontent.com/mahfouz72/afd57aa6e51ca161b7927acd959e5211/raw/b2b2571b42b8fa0f5e95ea546e9fa16e98948980/check.xml

Diff Regression patch config: https://gist.githubusercontent.com/mahfouz72/33967bf86e4a29cafa429771b0e95808/raw/761163c3fb766f95d7303e07c60ef178013be96e/check_patch.xml

Diff Regression projects: https://gist.githubusercontent.com/mahfouz72/a3d0af030c8f5efd0d8a39f2c14750bc/raw/a3424ad9b6f722e5de1a927e3d85e383101b2b93/projects-to-test-on.properties

### CODE REVIEW

**Comment by mahfouz72** [comment]
github, generate site

**Comment by mahfouz72** [comment]
github, generate report

**Comment by nrmancuso** [comment]
@mahfouz72 please provide an analysis of the reports, it is not clear to me why we lost violations on `this();` in a few places.

**Comment by mahfouz72** [comment]
> please provide an analysis of the reports, it is not clear to me why we lost violations on `this();` in a few places.

Please consider the last [report](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a41bb9e_2024112416/reports/diff/index.html) (always) , not the first one. In the first one we lost violation on `this ()` because I missed the token in the config

**Comment by nrmancuso** [comment]
>Please consider the last [report](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a41bb9e_2024112416/reports/diff/index.html) (always) ,

Not sure if this is a good policy, especially when we are generating several reports purposefully. If we have some bad/ invalid report, please leave a comment or edit your existing `Github, generate report` comment to let us know that there was some issue or whatever. Especially if I do not see any commits pushed, I consider any report that is here to be something I should review.

**Review by nrmancuso — APPROVED** [review]
Looks good!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/MethodParamPadCheck.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
 /**
  * <p>
  * Checks the padding between the identifier of a method definition,
- * constructor definition, method call, or constructor invocation;
+ * constructor definition, method call, constructor invocation, record or record pattern;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/MethodParamPadCheck.java:34]: `record or` => `record, or`

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*
+MethodParamPad
+allowLineBreaks = true
+option = (default)nospace
```
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:4]: Show other inputs with different property values.

Same with regression.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.whitespace.methodparampad;
+
+public class InputMethodParamPadCheckRecordPattern {
+
+    void test(Object obj) {
+
+        // violation below, ''(' is preceded with whitespace'
+        if (obj instanceof ColoredPoint (Point p, String s)) {
```
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:19]: Also show examples with space after `(`. Like https://checkstyle.org/property_types.html#PadOption has.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            case ColoredPoint (Point p, String s) -> {}
+            case Point(int x, int y) -> {}
+            default -> {}
+        }
+
+        boolean b = obj instanceof ColoredPoint
+                (Point p, String s); // ok, because allow line breaks
+
+        boolean c = obj instanceof ColoredPoint
+                (Point   // ok, because allow line breaks
```
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:45]: Also show with new line after `(`.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/MethodParamPadCheck.java (hunk 1 above)
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/MethodParamPadCheck.java]: done

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java (hunk 2 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:4]: done

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java (hunk 3 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:19]: done. but for clarity, the spaces after `(` is a concern for `ParenPad`.
 `MethodParamPad` is concerned with the space before ` ( `

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java (hunk 4 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/methodparampad/InputMethodParamPadCheckRecordPattern.java:45]: done

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
```
Error:  Java heap space -> [Help 1]
java.lang.OutOfMemoryError: Java heap space
    at sun.nio.cs.StreamEncoder.write (StreamEncoder.java:133)

```

**Comment by rnveach** [comment]
If regression prints too many violations then it will run out of memory. That is most likely the case here. I would remove the project(s) that cause the issues.

**Comment by mahfouz72** [comment]
Yes, I just change the `option` to `space` this definitely produces a huge amount of violation in all projects.
it was `no space` before

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a41bb9e_2024112416/reports/diff/index.html

**Comment by mahfouz72** [comment]
github, generate report

**Review by rnveach — APPROVED** [review]

**Pull request merged**